### ISSUE TITLE
Bootstrapper looses connection to che server

### ISSUE THREAD

**Issue by toorog** [eclipse-che/che#11147]
### Description
When provisioning a workspace I've observed the following behavior on web-ui:

1. Machines are reported as "Running"
2. Agents inside the machine fire up one after another
3. In my scenario one of these agents takes about 20s to be ready installed by the bootstrapper

Che never shows the agent as "Running". Instead the `/` circles forever. According to `/workspace_logs/bootstrapper/bootstrapper.log` the installation was successful.

Seems like che never gets the information that the agent installation succeeded. In the logs on the workspace you can see that the bootstrapper is trying to send an event via a closed channel. 

Any ideas how to solve this?

**OS and version:**    
Che 6.7.0 @ AKS

**Diagnostics:** 
```
2018/09/10 12:49:52 Bootstrapper configuration
2018/09/10 12:49:52   Push endpoint: wss://che.my-domain.com/api/websocket
2018/09/10 12:49:52   Push logs endpoint: wss://che.my-domain.com/api/websocket
2018/09/10 12:49:52   Auth enabled: true
2018/09/10 12:49:52   Runtime ID:
2018/09/10 12:49:52     Workspace: workspacet3u8f28vld9t09hb
2018/09/10 12:49:52     Environment: default
2018/09/10 12:49:52     OwnerId: 6b340c44-d2c9-4511-b663-115dda931beb
2018/09/10 12:49:52   Machine name: dev-machine
2018/09/10 12:49:52   Installer timeout: 180seconds
2018/09/10 12:49:52   Check servers period: 30seconds
2018/09/10 12:49:52   Push logs endpoint reconnect period: 10seconds
2018/09/10 12:49:52 Planning to install
2018/09/10 12:49:52 - org.eclipse.che.terminal:1.0.1 - Embedded web terminal
2018/09/10 12:49:52 - org.eclipse.che.ws-agent:1.0.3 - Workspace API support
2018/09/10 12:49:52 - org.eclipse.che.exec:1.0.1 - Agent for command execution
2018/09/10 12:49:52 - com.my-domain.custom.agent:1.0.0 - Agent that takes 20s to be installed
2018/09/10 12:49:52 Starting installations
2018/09/10 12:49:52 Installing 'org.eclipse.che.terminal'
2018/09/10 12:50:22 Installation of 'org.eclipse.che.terminal' successfully finished
2018/09/10 12:50:22 Trying to send event of type 'installer/statusChanged' to closed tunnel 'tunnel-1'
2018/09/10 12:50:22 Installing 'org.eclipse.che.ws-agent'
2018/09/10 12:50:22 Trying to send event of type 'installer/log' to closed tunnel 'tunnel-1'
2018/09/10 12:50:32 Successfully reconnected to logs endpoint
2018/09/10 12:50:52 Installation of 'org.eclipse.che.ws-agent' successfully finished
2018/09/10 12:50:52 Installing 'org.eclipse.che.exec'
2018/09/10 12:51:22 Installation of 'org.eclipse.che.exec' successfully finished
2018/09/10 12:51:22 Installing 'com.my-domain.custom.agent'
2018/09/10 12:51:34 Installation of 'com.my-domain.custom.agent' successfully finished
2018/09/10 12:51:34 All installations successfully finished
```

**Comment by sleshchenko** [comment]
@toorog Thanks for reporting this bug. As you said, for some reasons Che Server and Bootstrapper WebSocket connection was terminated. I suppose there can be the same issue with ingress domain restarting as we faced with nginx ingress controller (I've described this issue https://github.com/eclipse/che/issues/8675).
It is needed to check whether bootstrapper resends event after successful reconnecting to WebSocket endpoint and implement it if it is absent.

**Comment by toorog** [comment]
@sleshchenko I checked the bootstrapper go implementation already. Only a reconnect is done. The events won't be resent after reconnect succeeds.

**Comment by dorinburcea** [comment]
Hi @toorog I have same problem , not happen all the time and not all stacks.
My environment : 
che-server:6.9.0 in multiuser in k8s in AWS. 
Here how I reproduced:
1)worspace based on stack image registry.centos.org/che-stacks/centos-jdk8
2) Click on start the workspace 
3) In kubernetes the nginx ingress are created almost immediate
4) The workspace container is created and then the pod is running
5) I am able to navigate via chrome browser the at the ingress host that points to the ws-agent address http(s)://serverdr4fd7p5-sbweb3j-dev-machine-server-4401.<mydomain> 
6) Inside the workspace pod the the bootstraper log appear like
![image](https://user-images.githubusercontent.com/21106256/54343578-142da180-467a-11e9-9c53-55f8394e4b5f.png)
7) The web browser console log looks like this 
![Capture_websocket_console](https://user-images.githubusercontent.com/21106256/54343721-666ec280-467a-11e9-9e79-e3eee0331b90.PNG)
8) After a while dashboard is still not able to capture the workspace statusChange and workspace start timeout. The CHE clean-up process starts and remove the pods and all ingress. 

Thanks in advance for support

**Comment by PavelSosin** [comment]
Improvement proposal: bug or design issue? If workspace.NEXT has side-cars which run servers communicating with clients via tcp based protocol workspace has to track such side-cars state constantly. We had this problem when we implemented LSP server as CF application. In the case of WS protocol it was the real mess. We found simple solution: Old good keep-alive mechanism helped us to track LSP server application state at the client side. Is it obligatory for side-cars to fire kubernetes crash event to force pod restart?  Unfortunately, any TCP connection tends to be broken and server tends to crash in silent or simply answer 504 repeatedly due to running out of resources or bad request. Somebody has to keep air in the balloon warm. I suppose, tracking of side-cars state by workspace will be very helpful. Is it implemented or will be implemented in Che 7?
